core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+   */
+  public static Collection<Execution> getExecutionDirectEvents(EObject element, InstanceRole instanceRole) {
+    // get executions from cache
+    Collection<EObject> directEvents = getDirectEventsFromCache(element);
+    if (directEvents != null) {
+      return directEvents.stream().filter(Execution.class::isInstance).map(Execution.class::cast)
+          .collect(Collectors.toList());
+    }
+    // if not present, compute it and put it in cache
+    return getDirectEvents(element, instanceRole).stream().filter(Execution.class::isInstance)
+        .map(Execution.class::cast).collect(Collectors.toList());
+  }
+
+  /**
+   * Get StateFragment direct events on element.
+   * 
+   * @param element
+   *          EObject
+   * @param instanceRole
+   *          InstanceRole
+   * @return Collection<EObject>
+   */
+  public static List<StateFragment> getStateFragmentDirectEvents(EObject element, InstanceRole instanceRole) {
+    // get states from cache
+    Collection<EObject> directEvents = getDirectEventsFromCache(element);
+    if (directEvents != null) {
+      return directEvents.stream().filter(StateFragment.class::isInstance).map(StateFragment.class::cast)
+          .collect(Collectors.toList());
+    }
+    // if not present, compute it and put it in cache
+    return getDirectEvents(element, instanceRole).stream().filter(StateFragment.class::isInstance)
+        .map(StateFragment.class::cast).collect(Collectors.toList());
+  }
+
+  /**
+   * Compute EventContext structure for InstanceRole
+   * 
+   * @param instanceRole
+   *          InstanceRole
+   * @return List<EventContext>
+   */
+  public static List<EventContext> computeInstanceRoleEventContextStructure(InstanceRole instanceRole) {
+    if (instanceRole == null || !(instanceRole.eContainer() instanceof Scenario)) {
+      return Collections.emptyList();
+    }
+    Scenario scenario = SequenceDiagramServices.getScenario(instanceRole);
+
+    // initialize ancestors
+    Deque<CapellaElement> ancestors = new ArrayDeque<>();
+    ancestors.push(instanceRole);
+    List<EventContext> result = new ArrayList<>();
+
+    // Cache missing info from M2 :
+    // - cache Execution.start -> Execution and Execution.finish -> Execution when the end is a MessageEnd
+    // - cache StateFragment.start -> InteractionState and StateFragment.finish -> InteractionState
+    // (ExecutionEnd has a getExecution() method)(ExecutionEnd has a getExecution() method)
+    computeTimeLapseStartAndEndCache(scenario);
+
+    // compute Execution/StateFragment and InteractionFragment structure
+    Stream<InteractionFragment> ends = scenario.getOwnedInteractionFragments().stream()
+        .filter(frag -> frag instanceof AbstractEnd || frag instanceof InteractionState);
+    ends.forEachOrdered(end -> {
+      InstanceRole covered = getCoveredInstanceRole(end);
+      if (covered != null && covered.equals(instanceRole)) {
+        Optional<TimeLapse> timeLapse = getTimeLapseFromCache(end);
```

Comment:
check right away if `timelapse` is null / present here and thus avoid checking it in each of the visitMethods